### ISSUE TITLE
Webdav download with subdirectories creates only directories

### ISSUE THREAD

**Issue by joergbormet** [nextcloud/server#5543]
### Steps to reproduce
Dear nextcloud Team,please have a look on the following eMail dialog. For Reproduction feel free to use the readdle account on my nextcloud 12 Server installation. Best Regards Joerg 

Hello Joerg,

Thanks a lot for all the info provided, though our developers have confirmed, that the issue is occurring on the server side and we don't have any access to change this behavior.

Looks like the last update of the server dictates, that your files are folders, so the app draws the folder icon.

I may suggest you to contact your server administrator to address this issue.

Please contact me if you need any info and have an amazing week ahead.

Best Regards,
Olga Piskunovska
Readdle Support Team

Joerg Bormet - joerg.bormet@bormet.info - Jun 25 2017, 11:14 PM
Dear Olga,
for exact reproduction of the error I have created a user "readdle" (password: "Readdle#" (write it without hyphens)) on my nextcloud server. The URL is "https://nc.bormet.info/nextcloud/remote.php/webdav/". You will find a folder "a" with a subfolder "b" and a subsubfolder "c" with additional PDF-files within the folders. Start the Documents app. Connect to the cloud with the URL and the credentials given above and just download folder "a". Then have a look into the folders and subfolders. Only in folder "a" the PDF-file is downloaded als a pdf. In the subfolders the Documents app creates additional folders instead. I hope this will ultimately lead to the solution to fix this error.

Best regards
Joerg

Von meinem iPad gesendet

3.

### Expected behaviour
Tell us what should happen

### Actual behaviour
Tell us what happens instead

### Server configuration

**Operating system**:

**Web server:**

**Database:**

**PHP version:**

**Nextcloud version:** (see Nextcloud admin page)

**Updated from an older Nextcloud/ownCloud or fresh install:**

**Where did you install Nextcloud from:**

**Signing status:**
<details>
<summary>Signing status</summary>

```
Login as admin user into your Nextcloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results here.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
If you have access to your command line run e.g.:
sudo -u www-data php occ app:list
from within your Nextcloud installation folder
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
If you have access to your command line run e.g.:
sudo -u www-data php occ config:list system
from within your Nextcloud installation folder

or 

Insert your config.php content here. 
Make sure to remove all sensitive content such as passwords. (e.g. database password, passwordsalt, secret, smtp password, …)
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...

**Are you using encryption:** yes/no

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...

#### LDAP configuration (delete this part if not used)
<details>
<summary>LDAP config</summary>

```
With access to your command line run e.g.:
sudo -u www-data php occ ldap:show-config
from within your Nextcloud installation folder

Without access to your command line download the data/owncloud.db to your local
computer or access your SQL server remotely and run the select query:
SELECT * FROM `oc_appconfig` WHERE `appid` = 'user_ldap';


Eventually replace sensitive data as the name/IP-address of your LDAP server or groups.
```
</details>

### Client configuration
**Browser:**

**Operating system:**

#### Web server error log
<details>
<summary>Web server error log</summary>

```
Insert your webserver log here
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
Insert your Nextcloud log here
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
Insert your browser log here, this could for example include:

a) The javascript console log
b) The network log
c) ...
```
</details>

**Comment by rajo** [comment]
+1 I'm encountering exactly the same issue here; everything in the top level directory is synced perferctly but anything in a subdirectory is considered a directory even if it is a file thus breaking the document. Strangely this only happens if the directory is consiered a sync folder. If the folder sync is disabled in the documents app and one selects the nextcloud web share from the side panel directly, any file in any subdirectory can be accessed as it should. So I'm not entirely sure if it is solely nextcloud's failure here, as I as well cannot even view something in the access log when trying to access some file from a subdirectory in documents.

**Comment by FlyinFistOfJuda** [comment]
I have this problem too, tried it with Documents Version 5 and 6, same behaviour.

**Comment by critzinger** [comment]
i have the same problem see my post #5486

**Comment by FlyinFistOfJuda** [comment]
Hi @nickvergessen, sorry to bother, but is there any chance this issue may get into the 12.0.1 pipeline? It's quite a showstopper for workflows in environments where nextcloud is accessed via webdav clients.

**Comment by nickvergessen** [comment]
If someone posts step-by-step clear instructions how to reproduce, we may have a look

**Comment by joergbormet** [comment]
Sehr geehrter Herr Schilling, 
Sie können wie in meinem Post den Zugang von Readdle nutzen, den ich für Readdle erstellt hatte. Ich habe die im Post beschriebene Vorgehensweise mit meinem iPad Pro mit der aktuellen Dokuments—App durchgeführt. 
Liebe Grüße 
Jörg Bormet 

Am 6. Juli 2017 12:08:16 MESZ schrieb Joas Schilling <notifications@github.com>:
>If someone posts step-by-step clear instructions how to reproduce, we
>may have a look
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/nextcloud/server/issues/5543#issuecomment-313354596

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by FlyinFistOfJuda** [comment]
I do have a step by step instruction from my company that i have to anonymize first, this may take a while. If anyone else got one or is ready to create one, it may be quicker. 
@joergbormet : If the Devs opt in to test it on your server, the server may need to be running in debug level regarding rajo's comment of an empty access.log.

**Comment by rajo** [comment]
Hi there,

I've put some steps to recreate together (pdf with screenshots from the documents app). Can be found here: https://www.jochem.name/nextcloud/index.php/s/vbMe1OK0uUu9S7R

And I've enabled the debug mode on my nextcloud. As far as I can see, there occur many messages like the following when starting the documents-app-webdav-sync:

    rewinddir(): 88 is not a valid Directory resource at /var/www-ssl/nextcloud-12.0.0/3rdparty/icewind/streams/src/Wrapper.php#31

And in the apache logs, there seems to be no attempt to access some subdirectories by the app - just stuff from the top level directory is listed like this:

    "GET /nextcloud/remote.php/webdav/ebooks/smart_home2.pdf HTTP/1.1" 200 702197 "-" "ReaddleDAV Documents/6.0.2.354 (redirects)"

If anything further is needed - just ask

**Comment by rajo** [comment]
Been there, done that. They point fingers to nextcloud:

    Our specialists have checked the case and figured out that issue is occurring on the Nextcloud
    server side. Please note that as much as we'd love to help, there is nothing we can do on our side.
    The thing is that your server passes the information that your files are folders, so the app draws the
    folder icon. I suggest that you contact your server administrator to address this issue.

I can try again there though. I have too little knowledge of the webdav protocol to understand the expected behaviour and what kind of operations are supported and should work.

But: as I've written earlier, I have tested the very same documents app against an owncloud 9 instance and the sync worked! And the others above tell, that nextcloud versions < 12 also worked perfectly! So the claim that it is on their side doesn't seem to hold.

When diffing the nextcloud 11 and 12 versions, I can see that some pieces of code have changed in the webdav modules - although I am not in the position to tell what those changes are doing and if they could be relevant (but I wouldn't be surprised if they are ;) )

I was also so keen to try the webdav module of nc11 in nc12 but as it seems there have happened some api changes as well, so things don't fit easily and I have abandoned this. 

So in short again: I suspect some changes in the webdav subsytem between nc11 and nc12 but for myself I have no experience how to debug this. If someone can provide me with some instructions / proof-of-concept code, I am willing to try it on my server.

**Comment by nickvergessen** [comment]
> The thing is that your server passes the information that your files are folders, so the app draws the
folder icon.

Hmm okay, but that seems to be useful information.

**Comment by nickvergessen** [comment]
Can someone click that button and then look in the accesslog which path has been called? That would be very helpful. 👍

**Comment by nickvergessen** [comment]
Do you have a link to that conversion? Or was it by email?
Because I checked our dav backend, and it correctly reports folders as:
```xml
<d:resourcetype>
<d:collection/>
</d:resourcetype>
```

Files have no resource type but a mimetype as content type.
```xml
<d:getcontenttype>image/jpeg</d:getcontenttype>
<d:resourcetype/>
```

**Comment by rajo** [comment]
Conversation was by email.

**Comment by nickvergessen** [comment]
Can you forward it to https://github.com/nextcloud/server/blob/master/apps/dav/lib/CalDAV/Activity/Backend.php#L3 so I can reply back to the devs of the app and ask for their calls?

**Comment by rajo** [comment]
Sorry, now you've lost me :) forward what?

**Comment by nickvergessen** [comment]
The email of the documents app developers 😄

**Comment by critzinger** [comment]
The error is clearly with Nextcloud webdav server

I have the same problem with same answer from the manufacturer of app #5486

Until nextcloud 11.0.3 everything worked perfectly

**Comment by nickvergessen** [comment]
Yeah I emailed them, to find out what their request to the server is, let's see.

**Comment by FlyinFistOfJuda** [comment]
Just my 2 cent: Perhaps this only happens with mimetype:pdf ? I only experienced the problem with these files so far.

**Comment by rajo** [comment]
@FlyinFistOfJuda just tested that: created a subfolder with .doc .epub and .txt and .doc / .txt are shown as folders as well, .epub has an ?-icon and neither of them can be opened. Also makes no difference if there is only one file in the subfolder. Same error.

**Comment by blancjerome** [comment]
Hi,

I had the same problem with the brand new owncloud X and I found the way to correct it. Here is my correction submission, maybe it will help :)

https://github.com/owncloud/core/issues/28341

**Comment by rajo** [comment]
@blancjerome that's it! Although for nc12 the path is nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php 
I applied your proposed fixes and indeed, sync to the documents app is now working.

@nickvergessen if the proposed fix doesn't break something else, I would hereby vote to include it to the next release.

Thanks for everyone's help!

**Comment by critzinger** [comment]
@blancjerome thanks for your advice

i have tested it with nc 12.0.0 and pdf expert app

now it works perfectly again

@nickvergessen 
please includ the fix in the next release

**Comment by nickvergessen** [comment]
@LukasReschke patch dav in our 3rdparty until sabre-dav has it merged?

**Comment by nickvergessen** [comment]
@blancjerome do you want to report and fix the issue in https://github.com/fruux/sabre-dav ? Or shall I copy your solution and create a PR there?

**Comment by blancjerome** [comment]
You can report the fix to sabre-dav team.

Regards,

Jérôme

**Comment by nickvergessen** [comment]
Proposed the patch at https://github.com/fruux/sabre-dav/pull/982/files

I wrote a little test, which also confirms the issue and the solution. Thanks @blancjerome
